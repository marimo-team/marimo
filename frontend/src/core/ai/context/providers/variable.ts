/* Copyright 2024 Marimo. All rights reserved. */

import type { Completion } from "@codemirror/autocomplete";
import { createVariableInfoElement } from "@/core/codemirror/completion/variable-completions";
import type { DatasetTablesMap } from "@/core/datasets/data-source-connections";
import type { Variable, Variables } from "@/core/variables/types";
import { type AIContextItem, AIContextProvider } from "../registry";
import { Boosts } from "./common";

export interface VariableContextItem extends AIContextItem {
  type: "variable";
  data: {
    variable: Variable;
  };
}

export class VariableContextProvider extends AIContextProvider<VariableContextItem> {
  readonly title = "Variables";
  readonly mentionPrefix = "@";
  readonly contextType = "variable";

  constructor(
    private variables: Variables,
    private tablesMap: DatasetTablesMap,
  ) {
    super();
  }

  getItems(): VariableContextItem[] {
    const ignore = new Set(this.tablesMap.keys());

    return Object.entries(this.variables).flatMap(([name, variable]) => {
      if (ignore.has(name)) {
        return [];
      }
      return [
        {
          uri: this.asURI(name),
          name: name,
          type: this.contextType,
          description: variable.dataType ?? "",
          dataType: variable.dataType,
          data: {
            variable,
          },
        },
      ];
    });
  }

  formatCompletion(item: VariableContextItem): Completion {
    const { data } = item;
    const { variable } = data;
    return {
      label: `@${variable.name}`,
      displayLabel: variable.name,
      detail: variable.dataType ?? "",
      boost: Boosts.VARIABLE,
      type: this.contextType,
      section: "Variable",
      info: () => {
        return createVariableInfoElement(variable);
      },
    };
  }

  formatContext(item: VariableContextItem): string {
    const { uri: id, data } = item;
    const { variable } = data;
    return `Variable: ${id}\nType: ${variable.dataType || "unknown"}\nPreview: ${JSON.stringify(variable.value)}`;
  }

  getCompletions(): Completion[] {
    return [];
  }
}
