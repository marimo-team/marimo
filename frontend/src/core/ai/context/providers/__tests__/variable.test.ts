/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { describe, expect, it, vi } from "vitest";
import type { CellId } from "@/core/cells/ids";
import type { DatasetTablesMap } from "@/core/datasets/data-source-connections";
import type { Variable, VariableName, Variables } from "@/core/variables/types";
import { type VariableContextItem, VariableContextProvider } from "../variable";

// Mock the variable completions module
vi.mock("@/core/codemirror/completion/variable-completions", () => ({
  getVariableCompletions: vi.fn(
    (
      variables: Variables,
      _tableNames: string[],
      boost: number,
      prefix: string,
    ) => {
      return Object.entries(variables).map(([name, variable]) => ({
        name: `${prefix}${name}`,
        displayname: name,
        detail: variable.dataType || "unknown",
        boost,
        type: variable.dataType || "variable",
        apply: `${prefix}${name}`,
        section: "Variables",
      }));
    },
  ),
}));

// Mock data for testing
const createMockVariable = (
  name: string,
  options: Partial<Variable> = {},
): Variable => ({
  name: name as VariableName,
  declaredBy: ["cell1" as CellId],
  usedBy: ["cell2" as CellId],
  value: `value_of_${name}`,
  dataType: "str",
  ...options,
});

describe("VariableContextProvider", () => {
  describe("getItems", () => {
    it("should return empty array when no variables", () => {
      const variables: Variables = {};
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("empty-variables");
    });

    it("should return variable items for single variable", () => {
      const variables: Variables = {
        ["user_name" as VariableName]: createMockVariable("user_name", {
          value: '"John Doe"',
          dataType: "str",
        }),
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("single-variable");
    });

    it("should return variable items for multiple variables with different types", () => {
      const variables: Variables = {
        ["user_id" as VariableName]: createMockVariable("user_id", {
          value: "123",
          dataType: "int",
          declaredBy: ["cell1" as CellId],
          usedBy: ["cell2" as CellId, "cell3" as CellId],
        }),
        ["is_active" as VariableName]: createMockVariable("is_active", {
          value: "True",
          dataType: "bool",
          declaredBy: ["cell2" as CellId],
          usedBy: [],
        }),
        ["scores" as VariableName]: createMockVariable("scores", {
          value: "[1, 2, 3, 4, 5]",
          dataType: "list",
          declaredBy: ["cell3" as CellId],
          usedBy: ["cell4" as CellId],
        }),
        ["config" as VariableName]: createMockVariable("config", {
          value: '{"debug": true, "timeout": 30}',
          dataType: "dict",
          declaredBy: ["cell1" as CellId],
          usedBy: ["cell2" as CellId, "cell4" as CellId],
        }),
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("multiple-variables");
    });

    it("should handle variables with null/undefined values", () => {
      const variables: Variables = {
        ["null_var" as VariableName]: createMockVariable("null_var", {
          value: null,
          dataType: "NoneType",
        }),
        ["undefined_var" as VariableName]: createMockVariable("undefined_var", {
          value: undefined,
          dataType: null,
        }),
        ["empty_string" as VariableName]: createMockVariable("empty_string", {
          value: '""',
          dataType: "str",
        }),
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("null-undefined-variables");
    });

    it("should handle complex data types", () => {
      const variables: Variables = {
        ["df" as VariableName]: createMockVariable("df", {
          value: "<DataFrame shape: (100, 5)>",
          dataType: "pandas.DataFrame",
          declaredBy: ["cell1" as CellId],
          usedBy: ["cell2" as CellId, "cell3" as CellId],
        }),
        ["model" as VariableName]: createMockVariable("model", {
          value: "<sklearn.linear_model.LinearRegression>",
          dataType: "sklearn.linear_model._base.LinearRegression",
          declaredBy: ["cell4" as CellId],
          usedBy: ["cell5" as CellId],
        }),
        ["array" as VariableName]: createMockVariable("array", {
          value: "array([1, 2, 3, 4, 5])",
          dataType: "numpy.ndarray",
          declaredBy: ["cell2" as CellId],
          usedBy: ["cell3" as CellId],
        }),
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("complex-data-types");
    });

    it("should handle variables with special characters in names", () => {
      const variables: Variables = {
        ["_private_var" as VariableName]: createMockVariable("_private_var", {
          value: "42",
          dataType: "int",
        }),
        ["var_with_numbers123" as VariableName]: createMockVariable(
          "var_with_numbers123",
          {
            value: '"test"',
            dataType: "str",
          },
        ),
        ["CONSTANT_VAR" as VariableName]: createMockVariable("CONSTANT_VAR", {
          value: "3.14159",
          dataType: "float",
        }),
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();
      expect(items).toMatchSnapshot("special-chars-variables");
    });
  });

  describe("formatContext", () => {
    it("should format context for basic variable", () => {
      const variable = createMockVariable("username", {
        value: '"alice"',
        dataType: "str",
      });
      const item: VariableContextItem = {
        type: "variable",
        uri: "username",
        name: "username",
        description: "str",
        data: { variable },
      };

      const variables: Variables = { ["username" as VariableName]: variable };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("basic-variable-context");
    });

    it("should format context for variable without dataType", () => {
      const variable = createMockVariable("mystery_var", {
        value: "some_value",
        dataType: null,
      });
      const item: VariableContextItem = {
        type: "variable",
        uri: "mystery_var",
        name: "mystery_var",
        description: "",
        data: { variable },
      };

      const variables: Variables = {
        ["mystery_var" as VariableName]: variable,
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("no-datatype-variable-context");
    });

    it("should format context for variable with complex value", () => {
      const variable = createMockVariable("complex_data", {
        value:
          '{"users": [{"id": 1, "name": "Alice"}, {"id": 2, "name": "Bob"}], "total": 2}',
        dataType: "dict",
      });
      const item: VariableContextItem = {
        type: "variable",
        uri: "complex_data",
        name: "complex_data",
        description: "dict",
        data: { variable },
      };

      const variables: Variables = {
        ["complex_data" as VariableName]: variable,
      };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("complex-value-variable-context");
    });

    it("should format context for dataframe variable", () => {
      const variable = createMockVariable("sales_df", {
        value:
          "<DataFrame shape: (1000, 8)>\n   date    product  quantity  price\n0  2023-01-01  Widget A    10    29.99\n1  2023-01-02  Widget B     5    49.99\n...",
        dataType: "pandas.DataFrame",
      });
      const item: VariableContextItem = {
        type: "variable",
        uri: "sales_df",
        name: "sales_df",
        description: "pandas.DataFrame",
        data: { variable },
      };

      const variables: Variables = { ["sales_df" as VariableName]: variable };
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      const context = provider.formatContext(item);
      expect(context).toMatchSnapshot("dataframe-variable-context");
    });
  });

  describe("provider properties", () => {
    it("should have correct provider properties", () => {
      const variables: Variables = {};
      const tablesMap: DatasetTablesMap = new Map();
      const provider = new VariableContextProvider(variables, tablesMap);

      expect(provider.title).toBe("Variables");
      expect(provider.mentionPrefix).toBe("@");
      expect(provider.contextType).toBe("variable");
    });
  });

  describe("integration with tables", () => {
    it("should work with both variables and tables", () => {
      const variables: Variables = {
        ["df" as VariableName]: createMockVariable("df", {
          dataType: "pandas.DataFrame",
          value: "<DataFrame shape: (50, 3)>",
        }),
        ["connection_string" as VariableName]: createMockVariable(
          "connection_string",
          {
            dataType: "str",
            value: '"postgresql://localhost:5432/mydb"',
          },
        ),
      };

      const tablesMap: DatasetTablesMap = new Map([
        ["users", {} as any],
        ["products", {} as any],
      ]);

      const provider = new VariableContextProvider(variables, tablesMap);

      const items = provider.getItems();

      expect(items).toMatchSnapshot("variables-with-tables-items");
    });
  });
});
