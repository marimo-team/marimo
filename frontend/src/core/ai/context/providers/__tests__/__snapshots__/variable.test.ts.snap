// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VariableContextProvider > formatContext > should format context for basic variable > basic-variable-context 1`] = `
"Variable: username
Type: str
Preview: "\\"alice\\"""
`;

exports[`VariableContextProvider > formatContext > should format context for dataframe variable > dataframe-variable-context 1`] = `
"Variable: sales_df
Type: pandas.DataFrame
Preview: "<DataFrame shape: (1000, 8)>\\n   date    product  quantity  price\\n0  2023-01-01  Widget A    10    29.99\\n1  2023-01-02  Widget B     5    49.99\\n...""
`;

exports[`VariableContextProvider > formatContext > should format context for variable with complex value > complex-value-variable-context 1`] = `
"Variable: complex_data
Type: dict
Preview: "{\\"users\\": [{\\"id\\": 1, \\"name\\": \\"Alice\\"}, {\\"id\\": 2, \\"name\\": \\"Bob\\"}], \\"total\\": 2}""
`;

exports[`VariableContextProvider > formatContext > should format context for variable without dataType > no-datatype-variable-context 1`] = `
"Variable: mystery_var
Type: unknown
Preview: "some_value""
`;

exports[`VariableContextProvider > getItems > should handle complex data types > complex-data-types 1`] = `
[
  {
    "data": {
      "variable": {
        "dataType": "pandas.DataFrame",
        "declaredBy": [
          "cell1",
        ],
        "name": "df",
        "usedBy": [
          "cell2",
          "cell3",
        ],
        "value": "<DataFrame shape: (100, 5)>",
      },
    },
    "dataType": "pandas.DataFrame",
    "description": "pandas.DataFrame",
    "name": "df",
    "type": "variable",
    "uri": "variable://df",
  },
  {
    "data": {
      "variable": {
        "dataType": "sklearn.linear_model._base.LinearRegression",
        "declaredBy": [
          "cell4",
        ],
        "name": "model",
        "usedBy": [
          "cell5",
        ],
        "value": "<sklearn.linear_model.LinearRegression>",
      },
    },
    "dataType": "sklearn.linear_model._base.LinearRegression",
    "description": "sklearn.linear_model._base.LinearRegression",
    "name": "model",
    "type": "variable",
    "uri": "variable://model",
  },
  {
    "data": {
      "variable": {
        "dataType": "numpy.ndarray",
        "declaredBy": [
          "cell2",
        ],
        "name": "array",
        "usedBy": [
          "cell3",
        ],
        "value": "array([1, 2, 3, 4, 5])",
      },
    },
    "dataType": "numpy.ndarray",
    "description": "numpy.ndarray",
    "name": "array",
    "type": "variable",
    "uri": "variable://array",
  },
]
`;

exports[`VariableContextProvider > getItems > should handle variables with null/undefined values > null-undefined-variables 1`] = `
[
  {
    "data": {
      "variable": {
        "dataType": "NoneType",
        "declaredBy": [
          "cell1",
        ],
        "name": "null_var",
        "usedBy": [
          "cell2",
        ],
        "value": null,
      },
    },
    "dataType": "NoneType",
    "description": "NoneType",
    "name": "null_var",
    "type": "variable",
    "uri": "variable://null_var",
  },
  {
    "data": {
      "variable": {
        "dataType": null,
        "declaredBy": [
          "cell1",
        ],
        "name": "undefined_var",
        "usedBy": [
          "cell2",
        ],
        "value": undefined,
      },
    },
    "dataType": null,
    "description": "",
    "name": "undefined_var",
    "type": "variable",
    "uri": "variable://undefined_var",
  },
  {
    "data": {
      "variable": {
        "dataType": "str",
        "declaredBy": [
          "cell1",
        ],
        "name": "empty_string",
        "usedBy": [
          "cell2",
        ],
        "value": """",
      },
    },
    "dataType": "str",
    "description": "str",
    "name": "empty_string",
    "type": "variable",
    "uri": "variable://empty_string",
  },
]
`;

exports[`VariableContextProvider > getItems > should handle variables with special characters in names > special-chars-variables 1`] = `
[
  {
    "data": {
      "variable": {
        "dataType": "int",
        "declaredBy": [
          "cell1",
        ],
        "name": "_private_var",
        "usedBy": [
          "cell2",
        ],
        "value": "42",
      },
    },
    "dataType": "int",
    "description": "int",
    "name": "_private_var",
    "type": "variable",
    "uri": "variable://_private_var",
  },
  {
    "data": {
      "variable": {
        "dataType": "str",
        "declaredBy": [
          "cell1",
        ],
        "name": "var_with_numbers123",
        "usedBy": [
          "cell2",
        ],
        "value": ""test"",
      },
    },
    "dataType": "str",
    "description": "str",
    "name": "var_with_numbers123",
    "type": "variable",
    "uri": "variable://var_with_numbers123",
  },
  {
    "data": {
      "variable": {
        "dataType": "float",
        "declaredBy": [
          "cell1",
        ],
        "name": "CONSTANT_VAR",
        "usedBy": [
          "cell2",
        ],
        "value": "3.14159",
      },
    },
    "dataType": "float",
    "description": "float",
    "name": "CONSTANT_VAR",
    "type": "variable",
    "uri": "variable://CONSTANT_VAR",
  },
]
`;

exports[`VariableContextProvider > getItems > should return empty array when no variables > empty-variables 1`] = `[]`;

exports[`VariableContextProvider > getItems > should return variable items for multiple variables with different types > multiple-variables 1`] = `
[
  {
    "data": {
      "variable": {
        "dataType": "int",
        "declaredBy": [
          "cell1",
        ],
        "name": "user_id",
        "usedBy": [
          "cell2",
          "cell3",
        ],
        "value": "123",
      },
    },
    "dataType": "int",
    "description": "int",
    "name": "user_id",
    "type": "variable",
    "uri": "variable://user_id",
  },
  {
    "data": {
      "variable": {
        "dataType": "bool",
        "declaredBy": [
          "cell2",
        ],
        "name": "is_active",
        "usedBy": [],
        "value": "True",
      },
    },
    "dataType": "bool",
    "description": "bool",
    "name": "is_active",
    "type": "variable",
    "uri": "variable://is_active",
  },
  {
    "data": {
      "variable": {
        "dataType": "list",
        "declaredBy": [
          "cell3",
        ],
        "name": "scores",
        "usedBy": [
          "cell4",
        ],
        "value": "[1, 2, 3, 4, 5]",
      },
    },
    "dataType": "list",
    "description": "list",
    "name": "scores",
    "type": "variable",
    "uri": "variable://scores",
  },
  {
    "data": {
      "variable": {
        "dataType": "dict",
        "declaredBy": [
          "cell1",
        ],
        "name": "config",
        "usedBy": [
          "cell2",
          "cell4",
        ],
        "value": "{"debug": true, "timeout": 30}",
      },
    },
    "dataType": "dict",
    "description": "dict",
    "name": "config",
    "type": "variable",
    "uri": "variable://config",
  },
]
`;

exports[`VariableContextProvider > getItems > should return variable items for single variable > single-variable 1`] = `
[
  {
    "data": {
      "variable": {
        "dataType": "str",
        "declaredBy": [
          "cell1",
        ],
        "name": "user_name",
        "usedBy": [
          "cell2",
        ],
        "value": ""John Doe"",
      },
    },
    "dataType": "str",
    "description": "str",
    "name": "user_name",
    "type": "variable",
    "uri": "variable://user_name",
  },
]
`;

exports[`VariableContextProvider > integration with tables > should work with both variables and tables > variables-with-tables-items 1`] = `
[
  {
    "data": {
      "variable": {
        "dataType": "pandas.DataFrame",
        "declaredBy": [
          "cell1",
        ],
        "name": "df",
        "usedBy": [
          "cell2",
        ],
        "value": "<DataFrame shape: (50, 3)>",
      },
    },
    "dataType": "pandas.DataFrame",
    "description": "pandas.DataFrame",
    "name": "df",
    "type": "variable",
    "uri": "variable://df",
  },
  {
    "data": {
      "variable": {
        "dataType": "str",
        "declaredBy": [
          "cell1",
        ],
        "name": "connection_string",
        "usedBy": [
          "cell2",
        ],
        "value": ""postgresql://localhost:5432/mydb"",
      },
    },
    "dataType": "str",
    "description": "str",
    "name": "connection_string",
    "type": "variable",
    "uri": "variable://connection_string",
  },
]
`;
